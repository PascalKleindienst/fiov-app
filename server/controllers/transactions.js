// ========================================
// TRANSACTIONS CONTROLLER=================
// ========================================

// List all transactions of the current user.
module.exports.list_all_transactions = (req, res) => {
    req.app.locals.models.Transactions.findAll({ //findAndCountAll
        where: { user_id: req.user.id },
        include: ['category']
    })
        .then( (transactions) => {
            res.json(transactions);
        })
        .catch((err) => {
            res.send(err);
        });
};

// Create a new transaction for the current user.
module.exports.create_transaction = (req, res) => {
    const { description = '', note = '', amount = 0, createdAt = null } = req.body;

    req.app.locals.models.Transactions
        .create({ description, note, amount, createdAt, user_id: req.user.id })
        .then((transaction) => {
            res.json(transaction.get());
        })
        .catch((err) => {
            res.send(err);
        });
};
