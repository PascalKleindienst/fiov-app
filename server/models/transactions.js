// ========================================
// Transactions Model =====================
// ========================================
module.exports = (sequelize, DataTypes) => {
    const Transactions = sequelize.define('Transactions', {
        description: DataTypes.STRING,
        note: DataTypes.TEXT,
        amount: DataTypes.INTEGER
    }, {});

    Transactions.associate = function(models) {
        Transactions.belongsTo(models.Categories, {
            foreignKey: 'category_id',
            onDelete: 'CASCADE',
            as: 'category'
        });

        Transactions.belongsTo(models.User, {
            foreignKey: 'user_id',
            onDelete: 'CASCADE',
            as: 'user'
        });
    };
    
    return Transactions;
};