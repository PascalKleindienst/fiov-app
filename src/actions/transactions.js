// ==============================================
// TRANSACTIONS ACTIONS =========================
// ==============================================
import ajax from 'fetchival';

// =============== Add Transaction ===============

// =============== Remove Transaction ===============

// =============== Update Transaction ===============

// =============== Set Transactions ===============
export const SET_TRANSACTIONS = 'SET_TRANSACTIONS';

export const setTransactions = () => {
    return (dispatch) => {
        return ajax('/api/transactions', { mode: 'cors', credentials: 'same-origin' }).get().then((transactions) => {
            dispatch({
                type: SET_TRANSACTIONS,
                transactions
            });
        });
    };
};