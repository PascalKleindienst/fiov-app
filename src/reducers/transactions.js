// ==============================================
// TRANSACTIONS REDUCERS ========================
// ==============================================
import { SET_TRANSACTIONS } from '../actions/transactions';

const transactionsReducerDefaultState = [];

export default (state = transactionsReducerDefaultState, action) => {
    switch (action.type) {
    case SET_TRANSACTIONS:
        return action.transactions;
    default:
        return state;
    }
};
