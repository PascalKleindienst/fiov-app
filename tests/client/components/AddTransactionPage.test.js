// ==============================================
// ADD TRANSACTION PAGE TEST ====================
// ==============================================
// Mocks
const helpers = require('../../helpers');
import ajax from 'fetchival';
jest.mock('fetchival');

// Dependencies
import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import AddTransactionPageContainer, { AddTransactionPage } from '../../../src/components/AddTransactionPage';
import transactions from '../fixtures/transactions';

const mockStore = configureMockStore([thunk]);

describe('my AddTransactionPage component', () => {
    let wrapper, history, addTransaction;

    beforeEach(() => {
        addTransaction = jest.fn();
        history = { push: jest.fn() };
        wrapper = shallow(<AddTransactionPage addTransaction={addTransaction} history={history} translate={ (msg) => msg } />);
    });

    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should handle onSubmit', () => {
        wrapper.find('Connect(TransactionForm)').prop('onSubmit')(transactions[1]);

        expect(history.push).toHaveBeenLastCalledWith('/');
        expect(addTransaction).toHaveBeenLastCalledWith(transactions[1]);
    });

    test('should add transaction on form submit', () => {
        ajax.mockImplementation(() => ({ post: helpers.resolvePromise({ id: 1 }) }));
        const store = mockStore({
            locale: { languages: ['en'] }
        });
        const wrapper = shallow(<AddTransactionPageContainer store={store} />);
        
        return wrapper.props().addTransaction(transactions[1]).then(() => {
            const actions = store.getActions();
            expect(actions[0]).toEqual({ type: 'ADD_TRANSACTION', transaction: { ...transactions[1], id: 1} });
        });
    });
});
