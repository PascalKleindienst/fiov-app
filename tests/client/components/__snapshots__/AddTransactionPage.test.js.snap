// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my AddTransactionPage component should render correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        transactions.add_transaction
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(TransactionForm)
      onSubmit={[Function]}
    />
  </div>
</div>
`;
